Python/compile.c

```diff
@@ -6183,12 +6190,26 @@ compiler_pattern_mapping(struct compiler *c, pattern_ty p, pattern_context *pc)
             SET_LOC(c, ((pattern_ty) asdl_seq_GET(patterns, i)));
             return compiler_error(c, e);
         }
+
+        if (key->kind == Constant_kind) {
+            if (PySet_Contains(seen, key->v.Constant.value)) {
+                const char *e =  "Duplicate literal keys in match pattern are "
+                                 "not allowed";
+                return compiler_error(c, e);
```

Comment:
I think you need to call ``Py_DECREF(seen);`` too. The are also other 2 other ``compiler_error`` paths and you may need to decref too. Otherwise errors will leak memory.

If you want a cleaner implementation, maybe ``goto error`` and decref and ``return compiler_error(c, e);`` at that label. That way each branch only needs to set ``e`` then ``goto error``.

Reply:
Thank you, I really appreciate that feedback. See 3a96d32b94291ac9dcf5ba5cc9c09ed5a5f692f2 for the changes I made in response. I decided to not to use the `goto` pattern because different exit branches are doing different things. I could probably make it work with a `goto compiler_error` and `goto resource_error`, since those seem to be the two ways that this function exits, but then there is also the variadic call to `compiler_error` to think about if we use `goto`'s:

https://github.com/python/cpython/blob/082ded0e7672777e5ee0168cd1759a53b8c3f813/Python/compile.c#L6151-L6152

What do you think? Is it best to not use `goto`, all things considered?

Reply:
Sorry I wasn't being very clear. I mean't to consider using ``goto`` just for those 3 paths where you need to decref `seen`. Usually if there needs to be cleanup (like a decref), the convention is to goto ``error`` so that the cleanup code is grouped together and not all over the place. You can probably find this pattern in much of the C code in CPython.

However, since this code isn't too complex (yet). It's probably not needed. I'll just wait and see what the core devs have to say. Personally, I am fine with the code as-is.

Reply:
Please use `goto error;`. With my newest suggestions, I think that the amount of repeated error-handling code is becoming burdensome.

Your error block at the end of this function can just be:
```c
error:
    Py_DECREF(seen);
    return 0;
```

And then all of these error branches just become something like:

```c
const char *e = "...";
compiler_error(c, e);
goto error;
```